docs: update instructions to use Amplify UI components in app directory (#3850)
aws-amplify/amplify-ui: docs/src/pages/[platform]/getting-started/usage/nextjs/react.mdx

<InstallScripts framework="react" />

### Usage

<Tabs>
  <TabItem title="Next.js 13.4+">
  <Alert variation="info" role="none">
    Amplify UI components only work in Client Components. If you need to load them into a Server Component, you need to convert them into client-side components by adding a `'use client'` at the top of your file. Please see instructions below on how you could achieve this. For more info, visit Next.js [documentation](https://nextjs.org/docs/getting-started/react-essentials#third-party-packages).
  </Alert>
#### UI Components

##### Single Component

To use a Amplify UI component in Server Components, you must re-export them as Client Components.

An example below is shown for using Amplify UI `Button` component:

> First, create a Client Component that exports the Amplify component.  For example, create`AmplifyUIComponents.jsx` placing `'use client'` at the top , import the component and styles and re-export `Button` component.

Reviewer: Couple things here:
- prefer italicizing file names to give a create delineation between code itself and file names, same feedback for entire PR
- missing a space here: "create`AmplifyUIComponents.jsx`" 
```suggestion
First, create a Client Component that exports the Amplify component.  For example, create _AmplifyUIComponents.jsx_ placing `'use client'` at the top , import the component and styles and re-export `Button` component.
```

    <ExampleCode>

    ```tsx
    // AmplifyUIComponents.jsx
    'use client';

    import { Button } from '@aws-amplify/ui-react';

>     import '@aws-amplify/ui-react/styles.css';

Reviewer 2: Can you add a comment that if they are using Connected components, to put this in `configure.ts`?

Author: Maybe we should move the css to `configure-client`

    export { Button };
    ```

    </ExampleCode>

Then, import `Button` from the Client Component and use it anywhere you want, e.g., `app/page.jsx`.

    <ExampleCode>

    ```tsx
    // app/page.jsx
    import { Button } from './AmplifyUIComponents';

    export default function Home() {
      return <Button>Hello, Amplify!</Button>;
    }
    ```

    </ExampleCode>

> ##### Multiple Components

Author: Make sense

To use more Amplify UI components, you need to re-export them in the same Client Component `AmplifyUIComponents.jsx` and then load them from there.

    <ExampleCode>

    ```tsx
    // AmplifyUIComponents.jsx
    'use client';

    import { Alert, Button, Card } from '@aws-amplify/ui-react';
    import '@aws-amplify/ui-react/styles.css';

    export { Alert, Button, Card };
    ```

    </ExampleCode>

#### Connected Components

>     To use a Amplify UI Connected Component in app directory, you must set up `Amplify.configure` in a Client Component.

Reviewer: ```suggestion
To use an Amplify UI Connected Component in within the _app_ directory, you must run `Amplify.configure` in a Client Component file.
``

    First, create a Client Component, for example, `configure.js` by placing `'use client'` at the top of the file under `app` directory:

    <ExampleCode>

    ```tsx

>     // configure.js

Reviewer 2: What do you think of specifying that this is a client side configure for Amplify?
```suggestion
    // configure-client.js
```

Author: That makes sense

> 

Reviewer 2: ```suggestion
'use client';
```

Author: Thinking we should not limit the separate config to client. It could be running on the server as well. Wouldn't that give users an impression that it is client required only?

    import { Amplify } from "aws-amplify";

    import config from "../src/aws-exports";
    
    Amplify.configure(config);
    ```

    </ExampleCode>

>     Then import `configure.js` in `page.jsx` to finish the setup,

Reviewer 2: Update this to import into `layout.jsx`

   <ExampleCode>

    ```tsx

>     // page.jsx

Reviewer 2: update code to layout.jsx could probably just include top section since layout code is fairly long.

Author: Yes, just include the import statement

    import "./configure";

    export default function Home() {
      return <h1>Hello, Amplify!</h1>;
    }
    ```

    </ExampleCode>

> Like UI components, you must re-export the Connected Component as a Client Component before using them.

Reviewer 2: ```suggestion
Like UI components, you must re-export the Connected Component as a Client Component before using them in Server Components.
```

> An example below is shown for `Authenticator` component:

Reviewer 2: Let's update this example to StorageManager

> First, create a Client Component, for example, `AmplifyConnectedComponents.jsx` by placing `'use client'` at the top of the file under `app` directory,

Author: Make sense

For more info, see Next.js [documentation](https://nextjs.org/docs/getting-started/react-essentials#third-party-packages).

  </TabItem>
  <TabItem title="Next.js 11.0-13.3">
#### UI Components
To use Amplify UI components in page directory, simply import them into your modules.

An example below is shown for using Amplify UI `Button` component:

    <ExampleCode>

    ```tsx
    // src/pages/page.jsx
    import { Button } from '@aws-amplify/ui-react';
    import '@aws-amplify/ui-react/styles.css';

    export default function Page() {
      return (
       <Button>Hello, Amplify!</Button>
      );
    }
    ```

   </ExampleCode>

#### Connected Components
To use Amplify Connected componnets in page directory, you must set up `Amplify.configure` in `src/pages/_app.jsx`

An example below is shown for `Authenticator` component:

    <ExampleCode>

    ```tsx
    // src/pages/_app.jsx
    import { Amplify } from 'aws-amplify';

    import { Authenticator } from '@aws-amplify/ui-react';
    import '@aws-amplify/ui-react/styles.css';

    import awsExports from '../aws-exports';
    Amplify.configure(awsExports);

    export default function App({ Component, pageProps }) {

>       return (
>         // your app
>         <Authenticator><Component {...pageProps} /></Authenticator>
>       );

Reviewer 3: nit:
```suggestion
      return (
        // your app
        <Authenticator>
          <Component {...pageProps} />
        </Authenticator>
      );
```

<InstallScripts framework="react" />

### Usage

<Tabs>
  <TabItem title="Next.js 13.4+">
  <Alert variation="info" role="none">
    Amplify UI components only work in Client Components. If you need to load them into a Server Component, you need to convert them into client-side components by adding a `'use client'` at the top of your file. Please see instructions below on how you could achieve this. For more info, visit Next.js [documentation](https://nextjs.org/docs/getting-started/react-essentials#third-party-packages).
  </Alert>
#### UI Components
To use Amplify UI components in Server Components, you must re-export them as Client Components.

An example below is shown for using Amplify UI `Button` component:

> First, create a client module, for example, `AmplifyUIComponents.jsx` by placing `'use client'` at the top and re-export `Button` component.

Reviewer 4: ```suggestion
First, create a client component that exports the Amplify component.  For example, create`AmplifyUIComponents.jsx` placing `'use client'` at the top , import the component and styles and re-export `Button` component.
```

    <ExampleCode>

    ```tsx
    // AmplifyUIComponents.jsx
    'use client';

    import { Button } from '@aws-amplify/ui-react';
    import '@aws-amplify/ui-react/styles.css';

    export { Button };
    ```

    </ExampleCode>

Then, import `Button` from the client module and use it anywhere you want, e.g., `app/page.jsx`.

>     <Example>

Author: Good callout. I will do an audit to be consistent

    <ExampleCode>

    ```tsx
    // AmplifyConnectedComponents.jsx
    'use client';

    import { Authenticator } from '@aws-amplify/ui-react';
    import '@aws-amplify/ui-react/styles.css';
   
    export { Authenticator };
    ```

    </ExampleCode>

Then, import `Authenticator` from the client module and use it anywhere you want, e.g., `app/page.jsx`.

    <Example>
      <ExampleCode>

      ```tsx
      // app/page.jsx
      import { Authenticator } from './Authenticator';

      export default function Home() {
        return <Authenticator>Hello, Amplify!</Authenticator>;
      }
      ```

      </ExampleCode>
    </Example>

To use more Connected components, you need to re-export them in the same client module `AmplifyConnectedComponents.jsx` and then load them from there.

    <ExampleCode>

    ```tsx
    // AmplifyConnectedComponents.jsx
    'use client';

    import { Authenticator } from '@aws-amplify/ui-react';
    import { StorageManager } from '@aws-amplify/ui-react-storage';
    import '@aws-amplify/ui-react/styles.css';

    export { Authenticator, StorageManager };
    ```

    </ExampleCode>

> **You may want to combine the setup and re-exports into a single client module.**

Reviewer 5: TBH, I kinda like this way better. It seems weird to have a separate file to do the `Amplify.configure` and then not export anything? Should we just show this way? I don't think we need to show 2 options, its a bit verbose. I can also be wrong about this if you disagree.

Author: You are totally right and this my original opinion about how to configure. We went back and forth with this part and we got feedback from JS team that in cases where users would use SSR features, they need the config in a Server Component as well. So we might as well show a pattern to let users know, the config can be separated.

For example, create a client module `AmplifyUI.js` specific for Amplify UI under `app` directory and then finish the setup with `Amplify.configure` and re-export everything you need from there.

    <ExampleCode>

    ```tsx
    // AmplifyUI.js
    'use client';

    import { Amplify } from 'aws-amplify';

    import { Authenticator, Button } from '@aws-amplify/ui-react';
    import '@aws-amplify/ui-react/styles.css';

    import awsExports from '../src/aws-exports';
    Amplify.configure(awsExports);

    export { Authenticator, Button };
    ```

    </ExampleCode>

For more info, see Next.js [documentation](https://nextjs.org/docs/getting-started/react-essentials#third-party-packages).

  </TabItem>

>   <TabItem title="Next.js 11.0-13.3">

Author: Yeah, we recently updated this part I think. @wlee221 explained why <- https://github.com/aws-amplify/amplify-ui/pull/3826#discussion_r1185581271

<InstallScripts framework="react" />

### App Router
Next.js 13.4+ introduced the [App Router `/ app directory](https://nextjs.org/docs/app/building-your-application/routing#the-app-directory) with the usage of [React Server Components](https://nextjs.org/docs/getting-started/react-essentials#server-components). 

Amplify UI components are highly interactive and must be used as Client Components. In order to use Amplify UI components in a Server Component, you need to convert them to Client Components by adding a `'use client'` at the top of your file.

For more info, visit Next.js [third party package documentation](https://nextjs.org/docs/getting-started/react-essentials#third-party-packages).

#### UI Components

To use an Amplify UI component in Server Components, you must re-export them as Client Components.

An example below is shown for using Amplify UI `Button` component:

> First, create a Client Component file _AmplifyComponents.jsx_ that re-exports `Button`. Make sure to place `'use client'` at the top of _AmplifyComponents.jsx_ to let Nextjs know that should be handled as a Client Component.

Reviewer: ```suggestion
First, create a Client Component file _AmplifyComponents.jsx_ that re-exports `Button`. Make sure to place `'use client'` at the top of _AmplifyComponents.jsx_ to let Nextjs know that the file should be handled as a Client Component.
```

> First, create a Client Component file _AmplifyComponents.jsx_ that re-exports `Button`. Make sure to place `'use client'` at the top of _AmplifyConnectedComponents.jsx_ to let Nextjs know that should be handled as a Client Component.

Reviewer 2: ```suggestion
First, create a Client Component file _AmplifyComponents.jsx_ that re-exports `Button`. Make sure to place `'use client'` at the top of _AmplifyComponents.jsx_ to let Nextjs know that should be handled as a Client Component.
```

<ExampleCode>

```tsx
// AmplifyComponents.jsx
'use client';

// To allow client-based Amplify components to be rendered inside of React Server Components we need to:

// 1. Import the Amplify components to be used in the application
import { Button } from '@aws-amplify/ui-react';

// 2. Import the Amplify UI styles
import '@aws-amplify/ui-react/styles.css';

// 3. Re-export the Amplify components for use in the application
export { Button };
```

</ExampleCode>

Then, import `Button` from the Client Component and use it anywhere you want, e.g., `app/page.jsx`.

<ExampleCode>

```tsx
// app/page.jsx

> import { Button } from './AmplifyUIComponents';

Reviewer 2: ```suggestion
import { Button } from './AmplifyComponents';
```

    <BasicShoppingCard />
  </Example>

  </TabItem>
    <TabItem title="App Router (Next.js 13.4+)">
    <Alert variation="info" role="none">

>       Next.js 13.4+ introduced [app directory](https://nextjs.org/docs/app/building-your-application/routing#the-app-directory) with the usage of [Server Components](https://nextjs.org/docs/getting-started/react-essentials#server-components). For simplicity, in this guide, we placed `"use client"` at the top of `page.jsx` to mark it as a Client Component, but you could also wrap the UI components in your own Client Component.

Reviewer: Thinking the below could be somewhat confusing for users:
```
For simplicity, in this guide, we placed `"use client"` at the top of `page.jsx` to mark it as a Client Component, but you could also wrap the UI components in your own Client Component.
```

Maybe we just remove it?

### Advanced Demo

<Tabs>
  <TabItem title="Pages Router (Next.js 11.0+)">
If you’d like to play around with a more comprehensive demo of the Amplify UI library, create a separate file under the `pages` folder named `paintings.js`, and copy/paste this code into it:

    <ExampleCode>

```tsx file=./examples/paintings.ts
```

   </ExampleCode>

Then, replace the contents of `pages/index.js` with the code below, restart the app, and experience all the beautiful components in action!

    <Example>
      <ExampleCode>

```tsx file=./examples/AdvancedExampleCode.tsx
```

      </ExampleCode>
      <AdvancedShoppingCard />
    </Example>

  </TabItem>
  <TabItem title="App Router (Next.js 13.4+)">
  <Alert variation="info" role="none">

>     Next.js 13.4+ introduced [app directory](https://nextjs.org/docs/app/building-your-application/routing#the-app-directory) with the usage of [Server Components](https://nextjs.org/docs/getting-started/react-essentials#server-components). For simplicity, in this guide, we placed `"use client"` at the top of `page.jsx` to mark it as a Client Component, but you could also wrap the UI components in your own Client Component.

Reviewer 2: fixed

  </TabItem>
</Tabs>

> ### Connected Components

Reviewer 2: Rather than putting this callout at the bottom, should we put it at the top before the tutorial?

Reviewer 2: We should add an info Alert on the Authenticator page that links to this section

<Tabs>
  <TabItem title="Next.js 13.4+">

> To use Connected componnets in app directory, make sure you re-export `Authenticator` as a Client Component by placing "use client" at the top of the file and set up with `Amplify.configure`, for example:

Reviewer: This reads like re-exporting the `Authenticator` is going to allow for usage of any Connected Component?

Author: You are right... I am kinda biased when using Authenticator as an example

> To use Connected componnets in app directory, make sure you re-export `Authenticator` as a Client Component by placing "use client" at the top of the file and set up with `Amplify.configure`, for example:

Reviewer 2: ```suggestion
To use Amplify UI Connected Components in app directory, you must re-export `Authenticator` as a Client Component by placing "use client" at the top of the file. See [Next.js docs](https://nextjs.org/docs/getting-started/react-essentials#rendering-third-party-context-providers-in-server-components). 

An example below is shown for `Authenticator` component:
```

    <ExampleCode>

    ```tsx
    // Authenticator.jsx
    'use client';

    import { Amplify } from 'aws-amplify';

    import { Authenticator } from '@aws-amplify/ui-react';
    import '@aws-amplify/ui-react/styles.css';

    import awsExports from '../src/aws-exports';

>     Amplify.configure(awsExports);

Reviewer 3: `Amplify.configure` can be done serverside, right?

Author: It can. I think if people are using some Amplify SSR features, they need to run it on the server side as well. For us, we do not have anything specific on the server now.

Reviewer 3: Since this docs is just for UI components, do you think it makes sense to remove `Amplify.configure` from here so that we don't make an opinion on where customer should do it?
